File: server-php/src/Controllers/UploadsController.php
<?php

declare(strict_types=1);

namespace EnklWiki\Controllers;

use EnklWiki\Request;
use EnklWiki\Response;
use EnklWiki\Support\Dates;
use EnklWiki\Support\UploadRepository;
use PDO;

final class UploadsController
{
    private const CONTENT_TYPES = [
        'svg' => 'image/svg+xml', 'png' => 'image/png', 'jpg' => 'image/jpeg', 'jpeg' => 'image/jpeg',
        'mp3' => 'audio/mpeg', 'mp4' => 'video/mp4', 'pdf' => 'application/pdf',
    ];

    /** @param string[] $allowedExtensions */
    public function __construct(
        private readonly PDO $db,
        private readonly string $uploadsPath,
        private readonly array $allowedExtensions
    ) {
    }

    public function getAll(Request $request): Response
    {
        $rows = $this->db->query('SELECT * FROM uploads ORDER BY created_at DESC')->fetchAll();
        return Response::json(array_map(UploadRepository::toDto(...), $rows));
    }

    public function upload(Request $request): Response
    {
        $file = $request->files['file'] ?? null;
        if (!is_array($file) || ($file['error'] ?? UPLOAD_ERR_NO_FILE) !== UPLOAD_ERR_OK) {
            return Response::badRequest('No file was uploaded.');
        }
        if ((int) $file['size'] === 0) {
            return Response::badRequest('File is empty.');
        }

        $originalName = (string) $file['name'];
        $ext = strtolower(pathinfo($originalName, PATHINFO_EXTENSION));
        if (!in_array($ext, $this->allowedExtensions, true)) {
            return Response::badRequest(
                sprintf('".%s" files aren\'t supported here. Allowed: %s.', $ext, implode(', ', $this->allowedExtensions))
            );
        }

        if (!is_dir($this->uploadsPath)) {
            mkdir($this->uploadsPath, 0775, true);
        }

        $storedName = bin2hex(random_bytes(16)) . '.' . $ext;
        $destination = rtrim($this->uploadsPath, '/\\') . DIRECTORY_SEPARATOR . $storedName;

        $tmpName = (string) $file['tmp_name'];
        $moved = is_uploaded_file($tmpName) ? move_uploaded_file($tmpName, $destination) : copy($tmpName, $destination);
        if (!$moved) {
            return Response::serverError('Could not store the uploaded file.');
        }

        $contentType = self::CONTENT_TYPES[$ext] ?? 'application/octet-stream';
        $id = bin2hex(random_bytes(16));
        $now = Dates::nowIso();

        $stmt = $this->db->prepare(
            'INSERT INTO uploads (id, file_name, original_file_name, content_type, size, created_at)
             VALUES (:id, :fileName, :originalFileName, :contentType, :size, :createdAt)'
        );
        $stmt->execute([
            'id' => $id, 'fileName' => $storedName, 'originalFileName' => $originalName,
            'contentType' => $contentType, 'size' => (int) $file['size'], 'createdAt' => $now,
        ]);

        return Response::json([
            'id' => $id, 'fileName' => $storedName, 'originalFileName' => $originalName,
            'contentType' => $contentType, 'size' => (int) $file['size'], 'createdAt' => $now,
        ]);
    }

    public function download(Request $request): Response
    {
        $fileName = (string) $request->param('fileName');
        $stmt = $this->db->prepare('SELECT * FROM uploads WHERE file_name = :fileName');
        $stmt->execute(['fileName' => $fileName]);
        $upload = $stmt->fetch();
        if ($upload === false) {
            return Response::notFound();
        }

        // Defends against a file_name that (however it got there) contains
        // path-traversal segments — stored names are always our own
        // hex-plus-extension, but this keeps the guarantee explicit.
        $safeName = basename($upload['file_name']);
        $path = rtrim($this->uploadsPath, '/\\') . DIRECTORY_SEPARATOR . $safeName;
        if (!is_file($path)) {
            return Response::notFound();
        }

        return Response::file($path, $upload['content_type'], $upload['original_file_name']);
    }
}

File: server/EnklWiki.Api/Controllers/UploadsController.cs
using EnklWiki.Api.Data;
using EnklWiki.Api.Dtos;
using EnklWiki.Api.Models;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.StaticFiles;
using Microsoft.EntityFrameworkCore;

namespace EnklWiki.Api.Controllers;

[ApiController]
[Route("api/uploads")]
public class UploadsController(AppDbContext db, IConfiguration configuration) : ControllerBase
{
    private static readonly string[] AllowedExtensions = ["svg", "png", "jpg", "jpeg", "mp3", "mp4", "pdf"];
    private readonly FileExtensionContentTypeProvider _contentTypes = new();

    private string UploadsDirectory
    {
        get
        {
            var path = configuration["Uploads:Path"] ?? "uploads";
            Directory.CreateDirectory(path);
            return Path.GetFullPath(path);
        }
    }

    [HttpGet]
    public async Task<ActionResult<List<UploadDto>>> GetAll()
    {
        var uploads = await db.Uploads
            .OrderByDescending(u => u.CreatedAt)
            .Select(u => new UploadDto(u.Id, u.FileName, u.OriginalFileName, u.ContentType, u.Size, u.CreatedAt))
            .ToListAsync();
        return Ok(uploads);
    }

    [Authorize]
    [HttpPost]
    [RequestSizeLimit(50 * 1024 * 1024)]
    public async Task<ActionResult<UploadDto>> Upload(IFormFile file)
    {
        if (file.Length == 0) return BadRequest("File is empty.");

        var ext = Path.GetExtension(file.FileName).TrimStart('.').ToLowerInvariant();
        if (!AllowedExtensions.Contains(ext))
        {
            return BadRequest($"\".{ext}\" files aren't supported here. Allowed: {string.Join(", ", AllowedExtensions)}.");
        }

        var storedName = $"{Guid.NewGuid():N}.{ext}";
        var fullPath = Path.Combine(UploadsDirectory, storedName);
        await using (var stream = System.IO.File.Create(fullPath))
        {
            await file.CopyToAsync(stream);
        }

        if (!_contentTypes.TryGetContentType(storedName, out var contentType))
        {
            contentType = "application/octet-stream";
        }

        var upload = new Upload
        {
            Id = Guid.NewGuid().ToString(),
            FileName = storedName,
            OriginalFileName = file.FileName,
            ContentType = contentType,
            Size = file.Length,
            CreatedAt = DateTimeOffset.UtcNow
        };
        db.Uploads.Add(upload);
        await db.SaveChangesAsync();

        return Ok(new UploadDto(upload.Id, upload.FileName, upload.OriginalFileName, upload.ContentType, upload.Size, upload.CreatedAt));
    }

    [HttpGet("{fileName}")]
    public async Task<IActionResult> Download(string fileName)
    {
        var upload = await db.Uploads.FirstOrDefaultAsync(u => u.FileName == fileName);
        if (upload is null) return NotFound();

        var fullPath = Path.Combine(UploadsDirectory, upload.FileName);
        if (!System.IO.File.Exists(fullPath)) return NotFound();

        return PhysicalFile(fullPath, upload.ContentType, upload.OriginalFileName);
    }
}
